wasm/derive/src/lib.rs

```diff
@@ -162,68 +76,34 @@ impl syn::parse::Parse for ParamType {
 /// ```
 #[proc_macro_attribute]
 pub fn entrypoint(attr: TokenStream, item: TokenStream) -> TokenStream {
-    let syn::ItemFn {
-        attrs,
-        vis,
-        sig,
-        mut block,
-    } = parse_macro_input!(item);
-
-    assert!(
-        syn::ReturnType::Default == sig.output,
-        "Exported function must not have a return type"
-    );
-
-    let args = match syn::parse_macro_input!(attr as Attr) {
-        Attr::Params(param_attr) => construct_args(&param_attr.params.types),
-        Attr::Empty => Punctuated::new(),
-    };
-    let fn_name = &sig.ident;
-
-    block.stmts.insert(
-        0,
-        parse_quote!(
-            use iroha_wasm::Execute as _;
-        ),
-    );
-
-    quote! {
-        /// Smart contract entry point
-        #[no_mangle]
-        pub unsafe extern "C" fn _iroha_wasm_main(
-        ) {
-            #fn_name(#args)
-        }
-
-        #(#attrs)*
-        #vis #sig
-        #block
-    }
-    .into()
+    entrypoint::impl_entrypoint(attr, item)
 }
 
-fn construct_args(
-    types: &Punctuated<ParamType, syn::token::Comma>,
-) -> Punctuated<syn::Expr, syn::token::Comma> {
-    types
-        .iter()
-        .map(|param_type| -> syn::Expr {
-            match param_type {
-                ParamType::Authority => {
-                    parse_quote! {
-                        ::iroha_wasm::query_authority()
-                    }
-                }
-                ParamType::TriggeringEvent => {
-                    parse_quote! {{
-                        use ::iroha_wasm::debug::DebugExpectExt as _;
-
-                        let top_event = ::iroha_wasm::query_triggering_event();
-                        ::core::convert::TryInto::try_into(top_event)
-                            .dbg_expect("Failed to convert top-level event to the concrete one")
-                    }}
-                }
-            }
-        })
-        .collect()
+/// Annotate the user-defined function that starts the execution of a validator.
+///
+/// Annotated function should have one parameter of the type which implements
+/// `TryFrom<NeedsPermissionBox>`.
+///
+/// Validators are only checking if an operation is **invalid**, not its validness.
+/// A validator can either deny the operation or pass it to the next validator if there is one.
+///
+/// # Panics
+///
+/// - If the function does not have a return type
+///
+/// # Example
+///
+// `ignore` because this macro idiomatically should be imported from `iroha_wasm` crate.
+//
+/// ```ignore
+/// use iroha_wasm::validator::prelude::*;
+///
+/// #[entrypoint]
```

Comment:
No. Look at `wasm/src/lib.rs` `validator` mod and its reexports. I actually think, that it may be a good idea, to do such modules with preludes for all kind of smart contracts